apps/web/src/pages/templates/components/chat-editor/TemplateChatEditor.tsx

```diff
@@ -1,51 +1,76 @@
 import { LackIntegrationAlert } from '../LackIntegrationAlert';
-import { VariableManager } from '../VariableManager';
-import { TranslateProductLead } from '../TranslateProductLead';
+import { CustomCodeEditor } from '../CustomCodeEditor';
+import { ChatPreview } from '../../../../components/workflow/Preview';
+
+import { VariableManagerModal } from '../VariableManagerModal';
+import { VariableManagementButton } from '../VariableManagementButton';
 
 const templateFields = ['content'];
 
 export function TemplateChatEditor() {
-  const { readonly } = useEnvController();
+  const [showLoading, setShowLoading] = useState(false);
   const stepFormPath = useStepFormPath();
-  const stepFormErrors = useStepFormErrors();
   const { control } = useFormContext<IForm>();
   const variablesArray = useVariablesManager(templateFields);
   const { hasActiveIntegration } = useHasActiveIntegrations({
     channelType: ChannelTypeEnum.CHAT,
   });
+  const [variablesModalOpened, setVariablesModalOpen] = useState(false);
+
+  const { start, clear } = useTimeout(() => setShowLoading(false), 1000);
+  const handleContentChange = (value: string, onChange: (string) => void) => {
+    setShowLoading(true);
+    clear();
+    onChange(value);
+    start();
+  };
 
   return (
     <>
       {!hasActiveIntegration ? <LackIntegrationAlert channelType={ChannelTypeEnum.CHAT} /> : null}
       <StepSettings />
-      <Controller
-        name={`${stepFormPath}.template.content`}
-        defaultValue=""
-        control={control}
-        render={({ field }) => (
-          <Textarea
-            {...field}
-            data-test-id="chatNotificationContent"
-            error={stepFormErrors ? stepFormErrors?.template?.content?.message : undefined}
-            disabled={readonly}
-            minRows={4}
-            value={(field.value as string) || ''}
-            label="Chat message content"
-            placeholder="Add notification content here..."
+      <Grid>
+        <Grid.Col span={6}>
+          <Controller
+            name={`${stepFormPath}.template.content`}
+            defaultValue=""
+            control={control}
+            render={({ field }) => (
+              <Stack spacing={8}>
+                <VariableManagementButton
+                  openVariablesModal={() => {
+                    setVariablesModalOpen(true);
+                  }}
+                />
+                <CustomCodeEditor
+                  value={(field.value as string) || ''}
+                  onChange={(value) => {
+                    handleContentChange(value, field.onChange);
+                  }}
+                />
+              </Stack>
+            )}
           />
-        )}
+        </Grid.Col>
+        <Grid.Col span={6}>
+          <ChatPreview showLoading={showLoading} />
+        </Grid.Col>
+      </Grid>
+      <VariableManagerModal
+        open={variablesModalOpened}
+        setOpen={setVariablesModalOpen}
+        variablesArray={variablesArray}
```

Comment:
Can we move this to the `VariableManagementButton` as both are inter-dependent design wise.